### PR TITLE
Image preview outputs

### PR DESCRIPTION
**Pull request by joeyballentine**
Uses the new output broadcasting system to preview a new output type: LargeImageOutput. This completely replaces the ImagePreview component and allows a new node, _View Image_ to also show image outputs. There is also a new output component called DefaultImageOutput, which has a clickable button to create a new view node.

This also comes with a new bonus: Any node with no inputs with handles will automatically run once it is valid, and will re-run whenever their inputData changes. This means that certain nodes, like NCNN's _Load Model_ node, which were not previously able to be typed, are now typed correctly (and cached in the backend). This only applies to text outputs and directory outputs (as well as image outputs of course).

### Screenshots:

image
image

### CODE REVIEW

File: backend/src/nodes/utils/image_utils.py
Comment (joeyballentine) [inline backend/src/nodes/utils/image_utils.py:250]: We should decide on the webp PR before merging this

File: src/renderer/contexts/GlobalNodeState.tsx
Comment (joeyballentine) [inline src/renderer/contexts/GlobalNodeState.tsx:254]: Would it be a good idea to cache the outputData on refresh?

Comment (RunDevelopment) [reply src/renderer/contexts/GlobalNodeState.tsx:254]: IMO, no.

File: backend/src/nodes/utils/image_utils.py
Comment (RunDevelopment) [reply backend/src/nodes/utils/image_utils.py:250]: These feature as separate, so this PR should not include them. The difference between PNG and WEBP isn't that big anyway, so it's not like this change is required for this PR to work.

Comment (joeyballentine) [reply backend/src/nodes/utils/image_utils.py:250]: Right. I cherry picked this commit to make that PR and I just didn't remove it after. I figured we'd have merged it by now

File: src/renderer/components/node/NodeOutputs.tsx
Comment (RunDevelopment) [inline src/renderer/components/node/NodeOutputs.tsx:18]:
```suggestion
interface FullOutputProps extends Omit<Output, 'id' | 'type'>, OutputProps {
```

File: src/renderer/components/node/NodeOutputs.tsx
Comment (RunDevelopment) [inline src/renderer/components/node/NodeOutputs.tsx:54]: I don't think you need to treat this as anything special.

```suggestion
            OutputType = GenericOutput;
            break;
```

**Review by RunDevelopment — CHANGES_REQUESTED** [review]
I'm not done reviewing, but here's what I have so far.

File: src/renderer/components/outputs/LargeImageOutput.tsx
Comment (RunDevelopment) [inline src/renderer/components/outputs/LargeImageOutput.tsx:48]: This just doesn't work. If you create a simple chain with 2 Load Image nodes, and set an image for both of them, one will always be marked stale.

File: src/renderer/components/outputs/LargeImageOutput.tsx
Comment (RunDevelopment) [inline src/renderer/components/outputs/LargeImageOutput.tsx:119]: For how important this information is, the font is mighty small.
image

File: src/renderer/components/node/Node.tsx
Comment (RunDevelopment) [inline src/renderer/components/node/Node.tsx:186]: When changing whether a node is disabled or locked, the node will re-runs.

File: backend/src/process.py
Comment (joeyballentine) [reply backend/src/process.py:185]: This is so it doesn't broadcast when you run the chain. Without it, it gets from the cache and the broadcasts it again even though it doesn't need to, so you see the node flicker for a short while

Comment (RunDevelopment) [reply backend/src/process.py:185]: Then that's a frontend bug. The backend should be allowed to repeat broadcasts for nodes. Child nodes have problems otherwise.

I would suggest doing a deep equal on received broadcast data to figure out whether the broadcast data map needs to be updated at all.

Comment (joeyballentine) [reply backend/src/process.py:185]: How would child nodes have problems?

File: src/renderer/components/node/Node.tsx
Comment (joeyballentine) [reply src/renderer/components/node/Node.tsx:165]: This determines if a node should make a run/individual request. The conditions are:
1) it has no inputs that have handles (meaning they don't require input from other nodes, we can just send all the inputData we have)
2) the input data we have covers all non-optional inputs

I'm not really sure why I did that second one, to be honest. It might have just been something I was trying when I had an unrelated bug and I just never removed it. But theoretically it wouldn't be necessary to keep that since the validity checker already checks this, right?

File: src/renderer/components/outputs/LargeImageOutput.tsx
Comment (joeyballentine) [reply src/renderer/components/outputs/LargeImageOutput.tsx:48]: Oh yeah...... Got any ideas for how to do that better?

File: src/renderer/components/outputs/LargeImageOutput.tsx
Comment (joeyballentine) [reply src/renderer/components/outputs/LargeImageOutput.tsx:119]: It's also misaligned for you for some reason

File: backend/src/nodes/image_nodes.py
Comment (joeyballentine) [reply backend/src/nodes/image_nodes.py:212]: I was considering renaming preview image to "view image (external)", like it used to be before we removed the opencv preview

File: backend/src/process.py
Comment (joeyballentine) [reply backend/src/process.py:185]: The backend doesn't cache anything in the main cache for iterator children, if that's what you mean. Those get treated as an entirely new executor, remember?

I don't disagree I should do a deep equals check when setting stuff though, but I also think we shouldn't be wasting IPC time sending data we don't need to send.

File: src/renderer/components/outputs/LargeImageOutput.tsx
Comment (joeyballentine) [reply src/renderer/components/outputs/LargeImageOutput.tsx:48]: Ok, what i've done is just not show the stale indicator when it's a starting node. I think that's good enough

File: src/renderer/components/node/Node.tsx
Comment (joeyballentine) [reply src/renderer/components/node/Node.tsx]: Ah, I know what this was for now: When you clear a node, it detects that an input data change happened before it recomputes validity, and therefore it tries to run it again and errors

File: src/common/common-types.ts
Comment (RunDevelopment) [inline src/common/common-types.ts:65]: The context pane now needs to check whether outputs have handles to connect to. It needs to be added to the `getFirstPossibleOutput`, I think. Please add that check.

File: backend/src/run.py
Comment (RunDevelopment) [inline backend/src/run.py:198]: I don't like that all this executor code is now here. We should probably make a follow up PR adding model outputs. Then we could get rid of this, because we wouldn't need `get_extra_data` anymore.

Comment (RunDevelopment) [reply backend/src/run.py:198]: On that note: Is the `get_extra_data` in Load Image even used anymore?

File: src/renderer/components/node/Node.tsx
Comment (RunDevelopment) [inline src/renderer/components/node/Node.tsx:167]: This didn't fix the disabled bug I reported. Try repeated disabling and enabling a Load Image node.

File: src/renderer/components/outputs/DirectoryOutput.tsx
Comment (RunDevelopment) [inline src/renderer/components/outputs/DirectoryOutput.tsx:24]: I don't like this definition of a starting node. It's very implicit. Could you please make a function `isStartingNode(schema: NodeSchema)` (maybe somewhere in util, idk)? That way, you wouldn't need the comment either.

File: src/renderer/components/outputs/GenericOutput.tsx
Comment (RunDevelopment) [inline src/renderer/components/outputs/GenericOutput.tsx:14]: Pretty sure you can remove that now.

File: src/renderer/components/outputs/DirectoryOutput.tsx
Comment (RunDevelopment) [reply src/renderer/components/outputs/DirectoryOutput.tsx:24]: And please use the function everywhere, obviously.

File: src/renderer/components/outputs/OutputContainer.tsx
Comment (RunDevelopment) [inline src/renderer/components/outputs/OutputContainer.tsx:20]: What does "generic" even mean here?

File: src/renderer/contexts/ExecutionContext.tsx
Comment (RunDevelopment) [inline src/renderer/contexts/ExecutionContext.tsx:228]: To not read stale data (`outputDataMap`), move all of this into `setOutputDataMap`. React won't update a `useState` if the value return by the set-supplier function is that same as the input value.

File: src/renderer/contexts/ExecutionContext.tsx
Comment (RunDevelopment) [inline src/renderer/contexts/ExecutionContext.tsx:230]: These dependencies seem weird.

File: src/common/common-types.ts
Comment (joeyballentine) [reply src/common/common-types.ts:65]: Oh yeah, forgot about that.

File: backend/src/run.py
Comment (joeyballentine) [reply backend/src/run.py:198]: I've already finished the model outputs stuff locally. Just waiting for this to be finished before I make that PR so I can rebase. And no, that get extra data is no longer used. I'll remove it.

File: src/renderer/components/outputs/OutputContainer.tsx
Comment (joeyballentine) [reply src/renderer/components/outputs/OutputContainer.tsx:20]: It's basically whether or not to put the label/type props next to the handle vs above the input

File: src/renderer/contexts/ExecutionContext.tsx
Comment (joeyballentine) [reply src/renderer/contexts/ExecutionContext.tsx:230]: Yeah this is kinda a problem -- without putting something here that forces it to update (which is what I used unAnimate for here), it doesn't properly update. I have no idea why, but using outputDataMap doesnt just work.

File: src/renderer/components/node/Node.tsx
Comment (RunDevelopment) [inline src/renderer/components/node/Node.tsx:86]: I might have mentioned this before already, but it bares repeating: `as` should be a last resort.

```suggestion
    const [validity, setValidity] = useState<Validity>({
        isValid: false,
        reason: 'Validating nodes...',
    });
```

File: src/renderer/components/outputs/TextOutput.tsx
Comment (RunDevelopment) [inline src/renderer/components/outputs/TextOutput.tsx:10]: What's the reason for this component? The only difference between TextOutput, DirectoryOutput, and GenericOutput is how they handle broadcast data. They are identical otherwise.

File: src/renderer/contexts/ExecutionContext.tsx
Comment (RunDevelopment) [inline src/renderer/contexts/ExecutionContext.tsx:228]: ..... This is a deps array.

Comment (RunDevelopment) [reply src/renderer/contexts/ExecutionContext.tsx:228]: Also, why do you still depend on `outputDataMap` at all?

File: src/renderer/components/node/Node.tsx
Comment (RunDevelopment) [inline src/renderer/components/node/Node.tsx:172]: isStartingNode

**Review by RunDevelopment — CHANGES_REQUESTED** [review]
I think we are pretty much there. I just have a few more suggestions and nits.

File: src/renderer/components/outputs/OutputContainer.tsx
Comment (RunDevelopment) [reply src/renderer/components/outputs/OutputContainer.tsx:20]: Huh, the input container does the same thing. Strange, but alright.

File: src/renderer/contexts/ExecutionContext.tsx
Comment (joeyballentine) [reply src/renderer/contexts/ExecutionContext.tsx:228]: Try it yourself without doing this. It doesn't work.

File: src/renderer/components/outputs/TextOutput.tsx
Comment (joeyballentine) [reply src/renderer/components/outputs/TextOutput.tsx:10]: Exactly

File: src/renderer/contexts/ExecutionContext.tsx
Comment (RunDevelopment) [reply src/renderer/contexts/ExecutionContext.tsx:228]: You are creating a new object with render. So this hook constantly updates.

Comment (joeyballentine) [reply src/renderer/contexts/ExecutionContext.tsx:228]: Oh

Comment (joeyballentine) [reply src/renderer/contexts/ExecutionContext.tsx:228]: Well something is wrong with this regardless

File: src/renderer/components/outputs/TextOutput.tsx
Comment (RunDevelopment) [reply src/renderer/components/outputs/TextOutput.tsx:10]: "Exactly" what lol 

You just duplicated a bunch of code everywhere. Why didn't you just give GenericOutput a prop like `convertOutputToType?: (value: unknown) => Type | undefined`. Or to be more compatible with the whole pick component thingy, you could also have made a higher kind component that is generic over the prop I suggested. So e.g.
```ts
const OutputComponent = (convertOutputToType?: (value: unknown) => Type | undefined) => {
  return memo(({ ... }: OutputProps) => { 
    // code of GenericOutput, but handling broadcasts like TextOutput with convertOutputToType 
  })
}

const TextOutput = OutputComponent(toStringType);
const DirectoryOutput = OutputComponent(toDirectoryType);
const GenericOutput = OutputComponent();
```

Comment (joeyballentine) [reply src/renderer/components/outputs/TextOutput.tsx:10]: I like that idea. I'll do that.

Comment (joeyballentine) [reply src/renderer/components/outputs/TextOutput.tsx:10]: Ended up not doing it that way because I felt like it was too complicated and we already pass the kind into the full props, so might as well just handle it in the type setter function itself

File: src/renderer/components/node/NodeOutputs.tsx
Comment (RunDevelopment) [inline src/renderer/components/node/NodeOutputs.tsx:41]:
```suggestion
        case 'directory':
        case 'text':
        case 'generic':
            OutputType = GenericOutput;
            break;
```

File: src/renderer/components/outputs/GenericOutput.tsx
Comment (RunDevelopment) [inline src/renderer/components/outputs/GenericOutput.tsx:25]: The value might an empty string.

```suggestion
                if (value !== undefined) {
```

File: src/renderer/contexts/ExecutionContext.tsx
Comment (RunDevelopment) [reply src/renderer/contexts/ExecutionContext.tsx:228]: You what? Fuck it. I can't even fix it using my usual invisible `div` hack.

Just make the deps array `[{}, setOutputDataMap]`. This should make it obvious that something is wrong.

**Review by RunDevelopment — APPROVED** [review]

**Pull request merged**
